### PR TITLE
refactor [dk]linemanagers

### PR DESCRIPTION
**Pull request by slingamn**
* Synchronize map accesses
* Be more aggressive about cleaning up expired lines; this is now done in a separate goroutine, after the original read-only function call has returned
* CIDR dlines did not work (nil dereference of `addrInfo` instead of `netInfo`)

### CODE REVIEW

File: irc/dline.go [hunk 1]
```diff
@@ -167,30 +182,32 @@ func (dm *DLineManager) CheckIP(addr net.IP) (isBanned bool, info *IPBanInfo) {
...
-			continue
+	doCleanup := false
+	defer func() {
+		if doCleanup {
+			go func() {
+				dm.Lock()
+				defer dm.Unlock()
+				for key, netInfo := range dm.networks {
+					if netInfo.Info.Time.IsExpired() {
+						delete(dm.networks, key)
```
Comment (DanielOaks) [inline irc/dline.go]: Hmm, will deleting elements while we're iterating through it cause any issues? Same for the `entries` slice below in the kline manager

Comment (slingamn) [reply irc/dline.go]: This is apparently safe: https://stackoverflow.com/a/23230406

Comment (DanielOaks) [reply irc/dline.go]: Ah fair enough, awesome! All good in that case.

**Comment by DanielOaks** [comment]
Awesome, thanks for the PR!

**Pull request merged**